### PR TITLE
✨ Merkle Lib

### PR DESCRIPTION
**Pull request by rajivpo**
should be more gas efficient than the [OZ implementation](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/utils/cryptography/MerkleProof.sol) which was used as a starting point.

First time writing something only in Yul so any and all comments/suggestions are welcome 🙂

### CODE REVIEW

File: src/utils/MerkleProof.sol [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+        bytes32[] memory proof,
+        bytes32 root,
+        bytes32 leaf
+    ) internal pure returns (bool isValid) {
+         assembly {
+            let computedHash := leaf
+            let proofLength := mload(proof)
+
+            // exit early if empty proof is supplied
+            if iszero(proofLength) {
```
Comment (itsdevbear) [inline src/utils/MerkleProof.sol:15]: probably can remove this, adds gas for the success case.

**Comment by transmissions11** [comment]
this is dope! thanks @rajivpo 

would you mind running prettier and giving this some comments in the style of our other yul libs?

File: src/utils/MerkleProof.sol (hunk 1 above)
Comment (rajivpo) [reply src/utils/MerkleProof.sol:15]: this is true - @transmissions11 what do you think?

Comment (transmissions11) [reply src/utils/MerkleProof.sol:15]: As long as it still reverts in other cases I agree

**Comment by transmissions11** [comment]
Looking better! Do we have an estimate of how much gas this saves vs OZ? Would also like some fuzz tests… i believe there are some other repos we can reference for generating trees in solidity

**Comment by rajivpo** [comment]
> Looking better! Do we have an estimate of how much gas this saves vs OZ? Would also like some fuzz tests… i believe there are some other repos we can reference for generating trees in solidity

image

Can definitely add some fuzz tests too.

**Comment by transmissions11** [comment]
bump on fuzz tests, have a great use case for this 😉

**Comment by rajivpo** [comment]
> bump on fuzz tests, have a great use case for this 😉

ye will try to wrap up tomorrow

**Comment by transmissions11** [comment]
lfg

**Pull request merged**